src/ui/Frame.cpp

```diff
@@ -1197,4 +1245,121 @@ void Frame::activateEntry(entry_t& entry) {
 			(*entry.click)(args);
 		}
 	}
+}
+
+void createTestUI() {
+	Frame* window = gui->addFrame("window");
+	window->setSize(SDL_Rect{(Frame::virtualScreenX - 500) / 2, (Frame::virtualScreenY - 400) / 2, 500, 400});
+	window->setActualSize(SDL_Rect{0, 0, 1500, 1500});
+	window->setColor(SDL_MapRGBA(mainsurface->format, 128, 128, 160, 255));
+
+	{
+		Button* bt = window->addButton("closeButton");
+		bt->setBorder(3);
+		bt->setSize(SDL_Rect{10, 10, 50, 50});
+		bt->setText("x");
+		bt->setTooltip("Close window");
+		class Callback : public Widget::Callback {
+		public:
+			Callback(Frame& f) :
+				window(f) {}
+			virtual ~Callback() = default;
+			virtual int operator()(Widget::Args& args) const override {
+				window.removeSelf();
+				return 0;
+			}
+		private:
+			Frame& window;
+		};
+		bt->setCallback(new Callback(*window));
+	}
+
+	int y = 500;
+
+	{
+		Button* bt = window->addButton("testButton1");
+		bt->setBorder(3);
+		bt->setSize(SDL_Rect{510, y, 240, 50});
+		bt->setText("Normal button");
+		bt->setTooltip("Only pressed when button is held");
+
+		y += 60;
+	}
+
+	{
+		Button* bt = window->addButton("testButton2");
+		bt->setBorder(3);
+		bt->setSize(SDL_Rect{510, y, 240, 50});
+		bt->setText("Toggle button");
+		bt->setTooltip("Toggles on/off state");
+		bt->setStyle(Button::STYLE_TOGGLE);
+
+		y += 60;
```

Comment:
If we don't do this, will all of the buttons take up the same screen space?
Or is there a way to automatically figure out some of this for you? Like `y += bt->height()` or something, I guess...

Reply:
yeah, this is not declarative UI in the same way as HTML or something so each element still has to be placed manually (which has pros and cons). I do `y += 60` here so there's some padding between this element and the next one (I could have done `y += bt->getSize().h + 15`) not all the UI has to be defined exactly like this, that function is strictly a test case.